Hashtable.c

```diff
@@ -98,71 +92,91 @@ static unsigned int nextPrime(unsigned int n) {
    return n;
 }
 
-static HashtableItem* HashtableItem_new(hkey_t key, void* value) {
-   HashtableItem* this;
-
-   this = xMalloc(sizeof(HashtableItem));
-   this->key = key;
-   this->value = value;
-   this->next = NULL;
-   return this;
-}
-
 Hashtable* Hashtable_new(unsigned int size, bool owner) {
    Hashtable* this;
 
    this = xMalloc(sizeof(Hashtable));
    this->items = 0;
    this->size = size ? nextPrime(size) : 11;
-   this->buckets = (HashtableItem**) xCalloc(this->size, sizeof(HashtableItem*));
+   this->buckets = (HashtableItem*) xCalloc(this->size, sizeof(HashtableItem));
    this->owner = owner;
    assert(Hashtable_isConsistent(this));
    return this;
 }
 
 void Hashtable_delete(Hashtable* this) {
+
    assert(Hashtable_isConsistent(this));
-   for (unsigned int i = 0; i < this->size; i++) {
-      HashtableItem* walk = this->buckets[i];
-      while (walk != NULL) {
-         if (this->owner)
-            free(walk->value);
-         HashtableItem* savedWalk = walk;
-         walk = savedWalk->next;
-         free(savedWalk);
-      }
+
+   if (this->owner) {
+      for (unsigned int i = 0; i < this->size; i++)
+         free(this->buckets[i].value);
    }
+
    free(this->buckets);
    free(this);
 }
 
+static void insert(Hashtable* this, hkey_t key, void* value) {
+   unsigned int index = key % this->size;
+   unsigned int origIndex = index;
+   unsigned int probe = 0;
+
+   for (;;) {
+      if (!this->buckets[index].value) {
+         this->items++;
+         this->buckets[index].key = key;
+         this->buckets[index].probe = probe;
+         this->buckets[index].value = value;
+         return;
+      }
+
+      if (this->buckets[index].key == key) {
+         if (this->owner && this->buckets[index].value != value)
+            free(this->buckets[index].value);
+         this->buckets[index].value = value;
+         return;
+      }
+
+      if (probe > this->buckets[index].probe) {
+         HashtableItem tmp = this->buckets[index];
+
+         this->buckets[index].key = key;
+         this->buckets[index].probe = probe;
+         this->buckets[index].value = value;
+
+         key = tmp.key;
+         probe = tmp.probe;
+         value = tmp.value;
+      }
+
+      index = (index + 1) % this->size;
+      probe++;
+
+      assert(index != origIndex);
+   }
+}
+
 void Hashtable_setSize(Hashtable* this, unsigned int size) {
 
    assert(Hashtable_isConsistent(this));
 
-   unsigned int newSize = nextPrime(size);
-   HashtableItem** newBuckets = (HashtableItem**) xCalloc(newSize, sizeof(HashtableItem*));
+   HashtableItem* oldBuckets = this->buckets;
+   unsigned int oldSize = this->size;
+
+   this->size = nextPrime(size);
```

Comment:
Should validate that the new size fits all existing items.

As far as I understand the changed implementation we no longer have a chain of items attached per bucket, thus the number of items is strictly limited by the number of buckets.